include/ensmallen_bits/callbacks/progress_bar.hpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * @file progress_bar.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of a simple progress bar callback function.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_CALLBACKS_PROGRESS_BAR_HPP
+#define ENSMALLEN_CALLBACKS_PROGRESS_BAR_HPP
+
+namespace ens {
+
+/**
+ * A simple progress bar, based on the maximum number of optimizer iterations
+ * and the EndEpoch callback function.
+ */
+class ProgressBar
+{
+ public:
+  /**
+   * Set up the progress bar callback class with the given width and output
+   * stream.
+   *
+   * @param width Width of the bar.
+   * @param ostream Ostream which receives output from this object.
+   */
+  ProgressBar(const size_t width = 70,
+              std::ostream& output = arma::get_cout_stream()) :
+      step(100.0 / width),
+      output(output),
+      previousProgress(0)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Callback function called at the end of a pass over the data.
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param epoch The index of the current epoch.
+   * @param objective Objective value of the current point.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void EndEpoch(OptimizerType& optimizer,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                const size_t epoch,
+                const double /* objective */)
+  {
+    if (!optimizer->MaxIterations())
+    {
+      Warn << "Maximum number of iterations not defined (no limit),"
+           << " no progress bar shown." << std::endl;
+      return;
+    }
```

Comment:
Do you think it would be better to have one progress bar per epoch, like Keras does?  In this case we would need an additional `StepTaken()` callback available (or something like this).

Reply:
Do you think this could be an option? I think depending on the number of iterations the output could be usage. Also can you clarify the ``StepTake()`` idea, not sure I can follow.

Reply:
So, basically, if we want to have a per-epoch progress bar, we would need to know how far into the epoch we are.  If we had a `StepTaken()` callback that indicated when a step was taken, what the (partial?) objective was, how big the batch size was, etc., we could know how much of the epoch was done and be able to print the progress bar.

Reply:
Sorry for the slow response, I can get the objective from the last Evaluate call, but I don't see an easy way to get the batch size, unless the optimizer provides an option. Maybe I missed something.

Reply:
Right, sorry---the batch size doesn't make sense to pass, because it's optimizer-specific.  If the user wants to know the batch size (and they are writing the callback) they can just call `BatchSize()` from inside the callback.  For a progress bar we could do that (if a `BatchSize()` function exists and the optimizer is a separable function, which we can check with traits I guess).  Let me know what you think.

Reply:
Okay, just wanted to make sure I didn't missed anything.